File: week07/frequency.py
#!/usr/bin/python3

import glob, io, math, os, re, sys
from count_word import getWordCounts

__all__ = ["getWordFrequencies", "loadPoetFrequencies"]

def wordCountsToFrequencies(wordCounts, isLog = False):
    result = {}
    for word, count in wordCounts.items():
        if word == "_total":
            continue

        if isLog:
            result[word] = math.log((count + 1) / wordCounts["_total"])
        else:
            result[word] = count / wordCounts["_total"]

    if isLog:
        result["_zero"] = -math.log(wordCounts["_total"])

    return result

def loadPoetFrequencies(isLog = False, debugWord = False):
    result = {}
    for file in sorted(glob.glob("poets/*.txt")):
        with io.open(file, encoding = "utf8") as input:
            counts = getWordCounts(input)
            frequencies = wordCountsToFrequencies(counts, isLog)
            poet = re.sub("_", " ", os.path.splitext(os.path.basename(file))[0])
            result[poet] = frequencies

            if debugWord:
                if isLog:
                    print("log(({}+1)/{:6d}) = {:8.4f} {}".format(counts.get(debugWord, 0), counts["_total"], frequencies.get(debugWord, frequencies["_zero"]), poet))
                else:
                    print("{:4d}/{:6d} = {:.9f} {}".format(counts.get(debugWord, 0), counts["_total"], frequencies.get(debugWord, 0), poet))

    return result

if __name__ == "__main__":
    isLog = os.path.basename(__file__) == "log_probability.py"
    word = sys.argv[1].lower()
    loadPoetFrequencies(isLog, word)

File: week07/frequency.pl
#!/usr/bin/perl
use strict;
use warnings;

use English;
use File::Spec;

require "count_word.pl";

sub wordCountsToFrequencies {
    my ($wordCounts, $isLog) = @ARG;

    my %result;
    foreach my $word (keys %$wordCounts) {
        if ($word eq "_total") {
            next;
        }

        my $count = $wordCounts->{$word};
        if ($isLog) {
            $result{$word} = log(($count + 1) / $wordCounts->{"_total"});
        } else {
            $result{$word} = $count / $wordCounts->{"_total"};
        }
    }

    if ($isLog) {
        $result{"_zero"} = -log($wordCounts->{"_total"});
    }

    return %result;
}

sub loadPoetFrequencies {
    my ($isLog, $debugWord) = @ARG;

    my %result;
    foreach my $file (glob "poets/*.txt") {
        open my $input, "<", $file;
        my %counts = getWordCounts($input);
        my %frequencies = wordCountsToFrequencies(\%counts, $isLog);
        my $poet = (File::Spec->splitpath($file))[2];
        $poet =~ s/\.txt$//;
        $poet =~ s/_/ /g;
        $result{$poet} = \%frequencies;

        if ($debugWord) {
            if ($isLog) {
                printf("log((%d+1)/%6d) = %8.4f %s\n", $counts{$debugWord} || 0, $counts{"_total"}, $frequencies{$debugWord} || $frequencies{"_zero"}, $poet);
            } else {
                printf("%4d/%6d = %.9f %s\n", $counts{$debugWord} || 0, $counts{"_total"}, $frequencies{$debugWord} || 0, $poet);
            }
        }
    }

    return %result;
}

if (!caller) {
    my $isLog = (File::Spec->splitpath($PROGRAM_NAME))[2] eq "log_probability.pl";
    my $word = lc $ARGV[0];
    loadPoetFrequencies($isLog, $word);
}

1;
